### PR TITLE
Refresh Unity before running tests

### PR DESCRIPTION
**Pull request by van800**
https://github.com/JetBrains/resharper-unity/issues/831 and https://github.com/JetBrains/resharper-unity/issues/558

### CODE REVIEW

File: resharper/resharper-unity/src/Rider/UnitTesting/RunViaUnityEditorStrategy.cs [hunk 1]
```diff
@@ -107,31 +117,77 @@ Task IUnitTestRunStrategy.Run(IUnitTestRun run)
...
+            working = true;
+
+            mySolution.Locks.ExecuteOrQueueEx(run.Lifetime, "Wait EditorState != UnityEditorState.Refresh", () =>
+            {
+                mySolution.Locks.QueueRecurring(run.Lifetime, "Periodic wait EditorState != UnityEditorState.Refresh",
+                    TimeSpan.FromSeconds(1), () =>
+                    {
+                        if (myEditorProtocol.UnityModel.Value != null)
+                        {
+                            var rdTask = myEditorProtocol.UnityModel.Value.GetUnityEditorState.Start(RdVoid.Instance);
```
Comment (korifey) [inline resharper/resharper-unity/src/Rider/UnitTesting/RunViaUnityEditorStrategy.cs]: we should have extension method that you shouldn't write RdVoid.Instance

File: resharper/resharper-unity/src/Rider/UnitTesting/RunViaUnityEditorStrategy.cs [hunk 2]
```diff
@@ -107,31 +117,77 @@ Task IUnitTestRunStrategy.Run(IUnitTestRun run)
...
+                            rdTask?.Result.Advise(run.Lifetime, result =>
+                            {
+                                if (result.Result != UnityEditorState.Refresh)
+                                    working = false;
+                            });
+                        }
+                    });
+            });
+
+            JetDispatcher.RunOrSleep(() => !(!run.Lifetime.IsAlive || !working), TimeSpan.FromMinutes(30));
```
Comment (korifey) [inline resharper/resharper-unity/src/Rider/UnitTesting/RunViaUnityEditorStrategy.cs]: please use deMorgan laws

File: resharper/resharper-unity/src/Rider/UnitTesting/RunViaUnityEditorStrategy.cs [hunk 3]
```diff
@@ -107,31 +117,77 @@ Task IUnitTestRunStrategy.Run(IUnitTestRun run)
...
+            mySolution.Locks.ExecuteOrQueueEx(run.Lifetime, "RefreshBeforeUT", () =>
+            {
+                myRiderSolutionSaver.Save(run.Lifetime, mySolution, async () =>
+                {
+                    await myUnityRefresher.Refresh(false);
+                    working = false;
+                });
+            });
+
+            JetDispatcher.RunOrSleep(() => !(!run.Lifetime.IsAlive || !working), TimeSpan.FromMinutes(30));
```
Comment (korifey) [inline resharper/resharper-unity/src/Rider/UnitTesting/RunViaUnityEditorStrategy.cs]: why 30 mins?

File: resharper/resharper-unity/src/Rider/UnitTesting/RunViaUnityEditorStrategy.cs (hunk 1 above)
Comment (van800) [reply resharper/resharper-unity/src/Rider/UnitTesting/RunViaUnityEditorStrategy.cs]: I don't see such method.

**Comment by van800** [comment]
@ihachani finishing touches before merging

**Pull request merged**